### ISSUE TITLE
classes with user-written initializers allow param fields with default values that are not params.

### ISSUE THREAD

**Issue by cassella** [chapel-lang/chapel#8423]
### Summary of Problem

```chapel
config var x = 7;

class C {
  param p = x;
  proc init() { }
}

var c = new C();
writeln(c);
delete c;
```

Without the ``writeln(c)``, this compiles.  With it,

```
fortytwo@magrathea:~/src/chpl/tfn$ chpl --force-initializers wrparmem.chpl
In file included from /tmp/chpl-fortytwo-24984.deleteme/_main.c:28:0:
/tmp/chpl-fortytwo-24984.deleteme/ChapelIO.c: In function ‘writeThisFieldsDefaultImpl2’:
/tmp/chpl-fortytwo-24984.deleteme/ChapelIO.c:236:26: error: ‘tmp_chpl’ undeclared (first use in this function)
 writeIt_chpl(_ref_tmp_6, tmp_chpl);
                          ^
/tmp/chpl-fortytwo-24984.deleteme/ChapelIO.c:236:26: note: each undeclared identifier is reported only once for each function it appears in
/home/fortytwo/src/chapel/runtime/etc/Makefile.exe:27: recipe for target '/tmp/chpl-fortytwo-24984.deleteme/wrparmem.tmp' failed
make: *** [/tmp/chpl-fortytwo-24984.deleteme/wrparmem.tmp] Error 1
error: compiling generated source
```

Curiously, with the ``proc init() { }`` removed but with ``--force-initializers``, it reports

```
wrparmem.chpl:6: error: default value for param is not a param
```

which is what I expected to happen as written above.

**Associated Future Test(s):**

[`test/classes/initializers/generics/param-field-default-value-non-param-config-var.chpl`](
           https://github.com/chapel-lang/chapel/blob/master/test/classes/initializers/generics/param-field-default-value-non-param-config-var.chpl)

[edit 2/12 cassella to add path to future]

### Configuration Information

```
fortytwo@magrathea:~/src/chpl/tfn$ chpl --version
chpl version 1.17.0 pre-release (44fa845)
Copyright (c) 2004-2018, Cray Inc.  (See LICENSE file for more details)

fortytwo@magrathea:~/src/chpl/tfn$ printchplenv --anonymize
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: native *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: udp
  CHPL_GASNET_SEGMENT: everything
CHPL_TASKS: qthreads
CHPL_LAUNCHER: amudprun
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: intrinsics
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: gmp
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_AUX_FILESYS: none

fortytwo@magrathea:~/src/chpl/tfn$ gcc --version
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.6) 5.4.0 20160609
```

**Comment by lydia-duncan** [comment]
I suspect the default initializer is catching it via the default value for the corresponding argument it creates.  But it doesn't tell you the name of what triggered the error, so we can't tell that it isn't the exact symbol we were expecting.

**Comment by cassella** [comment]
In creating a future, I tried changing this one to use a non-param field instead of a config var, and it resulted in an internal error instead of the above behavior.  So I guess that'll be a new issue.

**Comment by benharsh** [comment]
I looked into this a little, and it's not clear to me how we should resolve this bug.

We end up doing something like this in the initializer:
```chpl
param tmp = x;
this.p = tmp;
```

I don't know why, but we currently disable some param checking for temps. Regardless, we don't want to emit an error for the thing named "tmp".

**Comment by cassella** [comment]
At @lydia-duncan's suggestion in developing the future for #8443, compiling with --verify changed the error.  It changes it here as well (and to the same error):

```
fortytwo@magrathea:~/src/chapel (master)$ chpl --verify test/classes/initializers/generics/param-field-default-value-non-param-config-var.chpl
$CHPL_HOME/modules/internal/ChapelIO.chpl:255: In function 'writeThisFieldsDefaultImpl':
$CHPL_HOME/modules/internal/ChapelIO.chpl:279: internal error: Refers to a local/arg in another function [checkAST.cpp:69]
```
